sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/MergeScalarSubqueries.scala

```diff
@@ -0,0 +1,422 @@
+ * :     +- Aggregate [b#234], [avg(a#233) AS avg(a)#236, sum(b#234) AS sum(b)#238L]
+ * :        +- Project [a#233, b#234]
+ * :           +- Relation default.t[a#233,b#234] parquet
+ * +- Project [scalarsubqueryreference(0, 0, DoubleType, 231) AS scalarsubquery()#241,
+ *             scalarsubqueryreference(0, 1, LongType, 232) AS scalarsubquery()#242L]
+ *    +- OneRowRelation
+ */
+object MergeScalarSubqueries extends Rule[LogicalPlan] with PredicateHelper {
+  def apply(plan: LogicalPlan): LogicalPlan = {
+    plan match {
+      case Subquery(_: CommonScalarSubqueries, _) => plan
+      case s: Subquery => s.copy(child = extractCommonScalarSubqueries(s.child))
+      case _: CommonScalarSubqueries => plan
+      case _ => extractCommonScalarSubqueries(plan)
+    }
+  }
+
+  /**
+   * An item in the cache of merged scalar subqueries.
+   *
+   * @param elements  List of attributes that form the scalar return value of a merged subquery
+   * @param plan      The plan of a merged scalar subquery
+   * @param merged    A flag to identify if this item is the result of merging subqueries.
+   *                  Please note that `elements.size == 1` doesn't always mean that the plan is not
+   *                  merged as there can be subqueries that are different ([[checkIdenticalPlans]]
+   *                  is false) due to an extra [[Project]] node in one of them. In that case
+   *                  `elements.size` remains 1 after merging, but the merged flag becomes true.
+   */
+  case class Header(elements: Seq[(String, Attribute)], plan: LogicalPlan, merged: Boolean)
+
+  private def extractCommonScalarSubqueries(plan: LogicalPlan) = {
+    val cache = ListBuffer.empty[Header]
+    val newPlan = removeReferences(insertReferences(plan, cache), cache)
+    if (cache.nonEmpty) {
+      val scalarSubqueries = cache.map {
+        case Header(elements, child, _) => ScalarSubquery(createProject(elements, child))
+      }.toSeq
+      CommonScalarSubqueries(scalarSubqueries, newPlan)
+    } else {
+      newPlan
+    }
+  }
+
+  // First traversal builds up the cache and inserts `ScalarSubqueryReference`s to the plan.
+  private def insertReferences(plan: LogicalPlan, cache: ListBuffer[Header]): LogicalPlan = {
+    plan.transformAllExpressionsWithPruning(_.containsAnyPattern(SCALAR_SUBQUERY)) {
+      case s: ScalarSubquery if s.children.isEmpty =>
+        val (subqueryIndex, headerIndex) = cacheSubquery(s.plan, cache)
+        ScalarSubqueryReference(subqueryIndex, headerIndex, s.dataType, s.exprId)
+    }
+  }
+
+  // Caching returns the index of the subquery in the cache and the index of scalar member in the
+  // "header", that `CreateNamedStruct(name1, attribute1, name2, attribute2, ...)` expression in a
+  // [[Project]] node.
+  private def cacheSubquery(plan: LogicalPlan, cache: ListBuffer[Header]): (Int, Int) = {
+    val output = plan.output.head
+    cache.zipWithIndex.collectFirst(Function.unlift { case (header, subqueryIndex) =>
```

Comment:
If `cache.zipWithIndex.collectFirst(Function.unlift {` is hard to read I can covert it to a loop.